tests/benchmark/utils/s2n_negotiate_client_benchmark.cc

```diff
@@ -0,0 +1,231 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <tests/benchmark/utils/s2n_negotiate_client_benchmark.h>
+#include <tests/benchmark/utils/shared_info.h>
+#include <benchmark/benchmark.h>
+
+extern "C" {
+#include "bin/common.h"
+#include <error/s2n_errno.h>
+
+#include "tls/s2n_connection.h"
+
+#include "stuffer/s2n_stuffer.h"
+
+}
+
+
+struct addrinfo hints, *ai_list, *ai;
+int add, sockfd = 0;
+const char *server_name = "localhost";
+const char *cipher_prefs = "test_all_tls12";
+s2n_status_request_type type = S2N_STATUS_REQUEST_NONE;
+char *psk_optarg_list[S2N_MAX_PSK_LIST_LENGTH];
+size_t psk_list_len = 0;
+
+static void setup_config() {
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+
+    if ((add = getaddrinfo(host, port, &hints, &ai_list)) != 0) {
+        fprintf(stderr, "getaddrinfo error: %s\n", gai_strerror(add));
+        printf("Error: %d\n", add);
+        printf("Errno: %s\n", strerror(errno));
+        exit(1);
+    }
+
+    int connected = 0;
+    while(connected == 0) {
+        for (ai = ai_list; ai != NULL; ai = ai->ai_next) {
+            if ((sockfd = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol)) == -1) {
+                continue;
+            }
+
+            if (connect(sockfd, ai->ai_addr, ai->ai_addrlen) == -1) {
+                close(sockfd);
+                continue;
+            }
+
+            connected = 1;
+            if (DEBUG_PRINT) {
+                printf("Connected to s2nd\n");
+            }
+            break;
+        }
+    }
+    return;
+}
+
+static void client_benchmark(benchmark::State& state, bool warmup) {
+    struct s2n_config *config = s2n_config_new();
+
+    struct verify_data *unsafe_verify_data = (verify_data *) malloc(sizeof(verify_data));;
+
+    if (config == NULL) {
+        print_s2n_error("Error getting new config");
+        exit(1);
+    }
+    if (DEBUG_PRINT) {
+        printf("Cipher preference = %s\n", cipher_prefs);
+    }
+
+    struct s2n_cipher_suite *cipher_suites_benchmark[] = {
+            all_suites[state.range(0)]
+    };
+
+    const struct s2n_cipher_preferences cipher_preferences_benchmark = {
+            s2n_array_len(cipher_suites_benchmark),
+            cipher_suites_benchmark,
+    };
+
+    const struct s2n_security_policy security_policy_benchmark = {
+            S2N_SSLv3,
+            &cipher_preferences_benchmark,
+            &kem_preferences_kms_pq_tls_1_0_2020_07,
+            &s2n_signature_preferences_20201021,
+            NULL,
+            &s2n_ecc_preferences_20201021,
+    };
+
+    config->security_policy = &security_policy_benchmark;
```

Comment:
I guess we're not using the public API, then?

Reply:
We want to try to capture as many different cipher suites as we can within the benchmark and found that this was the easiest way to do so. The benchmark also uses internal calls in many other places. For now, the benchmark is not intended to be used by customers and is placed within the tests/benchmark folder.